securedrop/journalist.py

```diff
@@ -333,6 +349,162 @@ def edit_account():
+    return source_tags
+
+
+def get_submission_tags(submission_id):
+    submission = Submission.query.filter(Submission.id == submission_id).first()
+    submission_tags = db_session.query(SubmissionTag, SubmissionLabelType) \
+                            .join(SubmissionLabelType) \
+                            .filter(SubmissionTag.submission_id == submission.id) \
+                            .group_by(SubmissionLabelType.label_text).all()
+    return submission_tags
+
+
+def delete_source_label(label_id):
+    matching_label = SourceLabelType.query.filter(SourceLabelType.id == label_id).all()
+    for label in matching_label:
+        db_session.delete(label)
+    db_session.commit()
+
+
+def delete_submission_label(label_id):
+    matching_label = SubmissionLabelType.query.filter(SubmissionLabelType.id == label_id).all()
+    for label in matching_label:
+        db_session.delete(label)
+    db_session.commit()
+
+
+@app.route('/admin/delete_source_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_source_label_type(tag_id):
+    delete_source_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_source_label_type/', methods=('POST',))
+@admin_required
+def admin_create_source_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_source_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SourceLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'source')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_submission_label_type/', methods=('POST',))
+@admin_required
+def admin_create_submission_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_submission_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SubmissionLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'submission')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/delete_submission_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_submission_label_type(tag_id):
+    delete_submission_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+def create_source_tag(source_id, label_id):
+    source = Source.query.filter(Source.id == source_id).first()
+    db_session.add(SourceTag(source_id=source.id,
+                             label_id=label_id))
+    db_session.commit()
+
+
+def delete_source_tag(source_id, label_id):
+    source = Source.query.filter(Source.id == source_id).first()
+    matching_tag = SourceTag.query.filter(and_(SourceTag.label_id == label_id,
+                                               SourceTag.source_id == source.id)).all()
```

Comment:
In general, these ORM queries read like queries written by an experienced SQL programmer (no offense!!) 😆 That being said, there is *probably* a more expressive and ORM-ish way to write this. In general, if we're going to use an ORM I think it's best to maximize its expressive potential.

This code is ok to merge, but I encourage you to research the ORM a bit more in depth to see if you can better leverage the ORM to make this query cleaner and more readable.